src/main/kotlin/reposfinder/logic/ReposFinder.kt

```diff
@@ -0,0 +1,189 @@
+package reposfinder.logic
+
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import reposfinder.api.GraphQLQueries
+import reposfinder.config.Config
+import reposfinder.filtering.Filter
+import reposfinder.utils.Logger
+import java.io.File
+import java.lang.Integer.max
+import java.util.Date
+
+/*
+ *  Input URLs format (exactly two slashes):
+ *  [...]/{OWNER}/{REPONAME}
+ */
+class ReposFinder(
+    private val config: Config
+) : Runnable {
+    private companion object {
+        const val TOKEN_SHOW_LENGTH = 20
+    }
+
+    enum class Status {
+        READY,
+        WORKING,
+        LIMITS_WAITING,
+        LIMITS_UPDATE_ERROR,
+        BAD_TOKEN_LIMITS,
+        INTERRUPTED,
+        ERROR,
+        DONE
+    }
+
+    @Volatile private var status = Status.READY
+
+    private val limits: RateLimits
+    private val logger: Logger
+
+    private val dumpDir = File(config.dumpDir)
+    val reposStorage: ReposStorage
+
+    private val jsonMapper = jacksonObjectMapper()
+
+    init {
+        dumpDir.mkdirs()
+        logger = Logger(config.logPath, isDebug = config.isDebug)
+        reposStorage = ReposStorage(config.urls, config.dumpDir, config.dumpEveryNRepos, logger = logger)
+        limits = RateLimits(config.isCore, config.isGraphQL, config.token, logger = logger)
+    }
+
+    override fun run() {
+        if (status != Status.READY) {
+            return
+        }
+        status = Status.WORKING
+        status = try {
+            logStartSummary()
+            limits.update()
+            logLimits()
+            searchRepos()
+            reposStorage.dump()
+            Status.DONE
+        } catch (e: Exception) {
+            logger.add("============== ERROR WHILE SEARCH RUNNING ==============")
+            logger.add(e.stackTraceToString())
+            Status.ERROR
+        } finally {
+            logEndSummary()
+            logger.dump()
+        }
+    }
+
+    fun interrupt() {
+        status = Status.INTERRUPTED
+    }
+
+    private fun searchRepos() {
+        if (limits.isNoLimits()) {
+            status = Status.BAD_TOKEN_LIMITS
+            logger.add("ZERO LIMITS ERROR (maybe bad token)")
+            return
+        }
+        for (repo in reposStorage.allRepos) {
+            waitingControl(limits.check())
+            if (status != Status.WORKING) {
+                logger.add("SEARCH WAS INTERRUPTED WITH STATUS $status")
+                break
+            }
+            logger.add("> processing: /${repo.owner}/${repo.name}")
+
+            val isCoreGood = repo.loadCore() && repo.isGood(config.coreFilters)
+            val isGraphQLGood = repo.loadGraphQL() && repo.isGood(config.graphQLFilters)
+
+            if (isCoreGood && isGraphQLGood) {
+                reposStorage.addGood(repo)
+            } else {
+                reposStorage.addBad(repo)
+            }
+        }
+    }
+
+    private fun Repository.loadCore(): Boolean {
```

Comment:
Why you didn't write all these functions inside the Repository class?

Reply:
fixed